File: go-backend-interview-prep/07_io_files_networking/10_dns_and_timeouts/main.go
package main

import (
	"context"
	"fmt"
	"io"
	"net"
	"net/http"
	"time"
)

func main() {
	// --- Example 1: DNS lookup ---
	fmt.Println("=== DNS Lookup ===")
	hosts := []string{"localhost", "example.com"}

	for _, host := range hosts {
		addrs, err := net.LookupHost(host)
		if err != nil {
			fmt.Printf("  %s: error: %v\n", host, err)
		} else {
			fmt.Printf("  %s -> %v\n", host, addrs)
		}
	}

	// --- Example 2: HTTP client timeout ---
	fmt.Println("\n=== HTTP Client Timeout ===")

	// Local server: one fast, one slow endpoint
	mux := http.NewServeMux()
	mux.HandleFunc("/fast", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte("fast response"))
	})
	mux.HandleFunc("/slow", func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(3 * time.Second)
		w.Write([]byte("slow response"))
	})

	ln, _ := net.Listen("tcp", "127.0.0.1:0")
	defer ln.Close()
	go http.Serve(ln, mux)
	base := "http://" + ln.Addr().String()

	// Fast request succeeds
	client := &http.Client{Timeout: 2 * time.Second}
	resp, err := client.Get(base + "/fast")
	if err != nil {
		fmt.Println("  fast: error:", err)
	} else {
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		fmt.Println("  fast:", string(body))
	}

	// Slow request times out
	_, err = client.Get(base + "/slow")
	if err != nil {
		fmt.Println("  slow: timed out (expected):", err)
	}

	// --- Example 3: context timeout per request ---
	fmt.Println("\n=== Context Timeout (per-request) ===")
	ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
	defer cancel()

	req, _ := http.NewRequestWithContext(ctx, "GET", base+"/slow", nil)
	noTimeoutClient := &http.Client{} // no global timeout
	_, err = noTimeoutClient.Do(req)
	if err != nil {
		fmt.Println("  context timeout (expected):", err)
	}

	// --- Example 4: dial timeout ---
	fmt.Println("\n=== Dial Timeout ===")
	// Try connecting to a non-routable address to trigger timeout
	start := time.Now()
	_, err = net.DialTimeout("tcp", "192.0.2.1:12345", 500*time.Millisecond)
	elapsed := time.Since(start)
	if err != nil {
		fmt.Printf("  dial timeout after %v: %v\n", elapsed.Round(time.Millisecond), err)
	}
}

File: go-backend-interview-prep/07_io_files_networking/10_dns_and_timeouts/main.py
"""DNS and timeouts -- Python equivalent of the Go example."""

import json
import socket
import threading
import time
import urllib.request
import urllib.error
from http.server import HTTPServer, BaseHTTPRequestHandler


class Handler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/fast":
            self.send_response(200)
            self.end_headers()
            self.wfile.write(b"fast response")
        elif self.path == "/slow":
            time.sleep(3)
            self.send_response(200)
            self.end_headers()
            self.wfile.write(b"slow response")

    def log_message(self, *args):
        pass


def main() -> None:
    # --- Example 1: DNS lookup ---
    print("=== DNS Lookup ===")
    hosts = ["localhost", "example.com"]
    for host in hosts:
        try:
            hostname, aliases, addrs = socket.gethostbyname_ex(host)
            print(f"  {host} -> {addrs}")
        except socket.gaierror as e:
            print(f"  {host}: error: {e}")

    # --- Example 2: HTTP client timeout ---
    print("\n=== HTTP Client Timeout ===")

    server = HTTPServer(("127.0.0.1", 0), Handler)
    port = server.server_address[1]
    base = f"http://127.0.0.1:{port}"
    threading.Thread(target=server.serve_forever, daemon=True).start()

    try:
        # Fast request
        resp = urllib.request.urlopen(f"{base}/fast", timeout=2)
        print(f"  fast: {resp.read().decode()}")
    except Exception as e:
        print(f"  fast: error: {e}")

    try:
        # Slow request times out
        urllib.request.urlopen(f"{base}/slow", timeout=2)
    except Exception as e:
        print(f"  slow: timed out (expected): {e}")

    # --- Example 3: socket connect timeout ---
    print("\n=== Socket Connect Timeout ===")
    start = time.perf_counter()
    try:
        # Non-routable address to trigger timeout
        socket.create_connection(("192.0.2.1", 12345), timeout=0.5)
    except (socket.timeout, OSError) as e:
        elapsed = time.perf_counter() - start
        print(f"  dial timeout after {elapsed*1000:.0f}ms: {e}")

    server.shutdown()


if __name__ == "__main__":
    main()
